### PR TITLE
Create a settings reference in the docs

### PR DESCRIPTION
**Pull request by sbc100**
This should be able to replace https://emsettings.surma.technology/

A big upside here is that we a link that we can use for each setting        
as well as the ability to link to a given setting internally. 

As part of this change I also updated the settings docs themselves to be
formatted in a ReSt-friendly way.

I also had to rename the section tag for `wasm_workers.rst` and             
`asyncify.rst` because the happen to collide with the ref for the                                                          
corresponding setting name.

Fixes: #16563

### CODE REVIEW

**Comment by RReverser** [comment]
cc @surma who will likely be interested

File: tools/maint/update_settings_docs.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+# Copyright 2021 The Emscripten Authors.  All rights reserved.
+# Emscripten is available under two separate licenses, the MIT license and the
+# University of Illinois/NCSA Open Source License.  Both these licenses can be
+# found in the LICENSE file.
+
+"""Builds the emscripten website from source and creates a new commit & branch
+in the emscripten-site repository containing the changes."""
```
Comment (dschuff) [inline tools/maint/update_settings_docs.py:8]: This comment is out of date.

Can we replace it with a little bit of documentation on the script, e.g. a sample import or how it recognizes tags?

File: tools/maint/update_settings_docs.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    'ENVIRONMENT_IS',
+    'stbi',
+    'emscripten_webgl_enable',
+    'library',
+  ]
+  for p in prefixes:
+    comment = comment.replace(p + '_*', p + '\\_*')
+  comment = comment.replace('`', '``')
+  comment = comment.replace('*', '\\*')
+  comment = comment.replace('"GL_"', '"GL\\_"')
```
Comment (dschuff) [inline tools/maint/update_settings_docs.py:63]: I guess GL can just go in the prefixes list above?

**Review by dschuff — APPROVED** [review]

File: tools/maint/update_settings_docs.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+to emscripten via ``-s`` on the command line.  For example
+``-sASSERTIONS`` or ``-sASSERTIONS=0``.  For more details see the
+:ref:`emcc <emcc-s-option-value>` documentation.
+
+.. Auto-generated by update_settings_docs.py. **DO NOT EDIT**
+'''
+
+all_tags = {
+  'link': 'Only applicable during linking',
+  'compile+link': None,
```
[the remark this answers is not in the document]
Comment (sbc100) [reply tools/maint/update_settings_docs.py]: Yes, I updated to the we mention is, and default to "linking only", which is documented at the top of this page.

File: tools/maint/update_settings_docs.py (hunk 2 above)
Comment (sbc100) [reply tools/maint/update_settings_docs.py:63]: I decided to go with modifying the existing comments rather than these magic replacements.

Comment (dschuff) [reply tools/maint/update_settings_docs.py]: That makes sense to me. What will happen when someone inevitably forgets to format a comment in settings.js the right way? Is there some part of the comment that we consider essential,  so we can check whether update_settings successfully parsed that?

Comment (dschuff) [reply tools/maint/update_settings_docs.py]: (also, do we still even need the double-backticks replacement?)

Comment (sbc100) [reply tools/maint/update_settings_docs.py]: If you mess up the syntax then the docs would fail to build

Comment (sbc100) [reply tools/maint/update_settings_docs.py]: Yes, all these replacements are gone now.

File: site/source/docs/tools_reference/settings_reference.rst [hunk 4]
```diff
@@ -0,0 +1,2376 @@
...
+=====================================
+
+Internal (testing only): Disables the blitOffscreenFramebuffer VAO path.
+
+TEST_MEMORY_GROWTH_FAILS
+========================
+
+Internal (testing only): Forces memory growing to fail.
+
+LEGACY_SETTINGS
```
Comment (kripken) [inline site/source/docs/tools_reference/settings_reference.rst:2362]: This line and onwards seems unnecessary?

File: tools/maint/update_settings_docs.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
+# Copyright 2021 The Emscripten Authors.  All rights reserved.
+# Emscripten is available under two separate licenses, the MIT license and the
+# University of Illinois/NCSA Open Source License.  Both these licenses can be
+# found in the LICENSE file.
+
+"""Convert src/settings.js into ReSt docs that get published as
+part of the emscripten docs.
+
+For paser for src/settings.js is somewhat fragile, and the
```
Comment (kripken) [inline tools/maint/update_settings_docs.py:10]: I don't understand the first two words here?

File: tools/maint/update_settings_docs.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+comments need to be written in a ReSt friendly way.  This
+means, for example, using double backticks for keywords and
+leaving a blank line before and after list blocks.
+
+The parser goes understand several "tags" which we use
```
Comment (kripken) [inline tools/maint/update_settings_docs.py:15]:
```suggestion
The parser does understand several "tags" which we use
```

File: tools/maint/update_settings_docs.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+from tools.utils import path_from_root, read_file, safe_ensure_dirs
+
+header = '''\
+.. _settings-reference:
+
+============================
+Emscripten Compiler Settings
+============================
+
+The following an complete list of settings that can be passed
```
Comment (kripken) [inline tools/maint/update_settings_docs.py:37]:
```suggestion
The following a complete list of settings that can be passed
```

File: tools/maint/update_settings_docs.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+and have no effect during compilation.
+
+.. Auto-generated by update_settings_docs.py. **DO NOT EDIT**
+'''
+
+all_tags = {
+  'link': '',
+  'compile+link': 'Applicable during both linking and compilation',
+  'compile': 'Only applicable during compilation',
+  'experimental': 'This is an exeprimental setting',
```
Comment (kripken) [inline tools/maint/update_settings_docs.py:52]:
```suggestion
  'experimental': 'This is an experimental setting',
```

**Review by kripken — COMMENTED** [review]
Great idea!

File: site/source/docs/tools_reference/settings_reference.rst (hunk 4 above)
Comment (sbc100) [reply site/source/docs/tools_reference/settings_reference.rst:2362]: Done

File: site/source/docs/porting/asyncify.rst [hunk 9]
```diff
@@ -1,8 +1,8 @@
-.. _Asyncify:
+.. _asyncify section:
```
Comment (kripken) [inline site/source/docs/porting/asyncify.rst:1]: Unrelated change? Also, does this work with a space in the name?

**Review by kripken — COMMENTED** [review]
lgtm % comment

File: site/source/docs/porting/asyncify.rst (hunk 9 above)
Comment (sbc100) [reply site/source/docs/porting/asyncify.rst:1]: Spaces are fine in ref names yes.


I put the reason for this in the PR description: I also had to rename the section tag for wasm_workers.rst and
asyncify.rst because the happen to collide with the ref for the
corresponding setting name.
